src/python/pants/core/goals/tailor.py

```diff
@@ -145,8 +158,9 @@ def for_target_type(
                 )
             )
 
-        default_sources = default_sources_for_target_type(target_type)
-        if (explicit_sources or triggering_sources) and not default_sources:
+        if (explicit_sources or triggering_sources) and not has_source_or_sources_field(
+            target_type
+        ):
```

Comment:
The reason this was necessary to change is that `default_sources_for_target_type` checks whether a source(s) field have default values. For the `resource` target, naturally, there aren't any defaults, however, it does have the `source` field. This code has always run only for the targets with defaults set for their source(s) field, so it worked.

Now, we can't do this any longer so I am just checking whether the source(s) field is there as it should be enough (and more closely matching what the assertion error message actually tells.

Reply:
For extra context, this is the reason that function call is there https://github.com/pantsbuild/pants/pull/12409/files